File: Algorithms/Sorting/Insertion Sort - Part 1/solution.js
//https://www.hackerrank.com/challenges/insertionsort1/problem
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

function insertionSort1(n, arr) {
    var insertValue = arr[n - 1];
    var j = n - 2;

    while(j >= 0){
        if(insertValue < arr[j])
            arr[j + 1] = arr[j];
        else
            break;

        j--;
        console.log(arr.join(" "));
    }

    arr[j + 1] = insertValue;
    console.log(arr.join(" "));

}

function main() {
    const n = parseInt(readLine(), 10);

    const arr = readLine().split(' ').map(arrTemp => parseInt(arrTemp, 10));

    insertionSort1(n, arr);
}

File: Algorithms/Sorting/Insertion Sort - Part 1/solution.cpp
//https://www.hackerrank.com/challenges/insertionsort1/problem
#include <bits/stdc++.h>

using namespace std;

vector<string> split_string(string);

void print(std::vector<int> arr){
    for(std::vector<int>::iterator i = arr.begin(); i != arr.end(); ++i){
        std::cout << *i << " ";
    }
    std::cout << std::endl;
}

void insertionSort1(int n, vector<int> arr) {
    int insertValue = arr.back();
    int j = n - 2;
    
    while(j >= 0){
        if(insertValue < arr[j])
            arr[j + 1] = arr[j];
        else
            break;

        j--;
        print(arr);
    }

    arr[j + 1] = insertValue;
    print(arr);
}

int main()
{
    int n;
    cin >> n;
    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    string arr_temp_temp;
    getline(cin, arr_temp_temp);

    vector<string> arr_temp = split_string(arr_temp_temp);

    vector<int> arr(n);

    for (int i = 0; i < n; i++) {
        int arr_item = stoi(arr_temp[i]);

        arr[i] = arr_item;
    }

    insertionSort1(n, arr);

    return 0;
}

vector<string> split_string(string input_string) {
    string::iterator new_end = unique(input_string.begin(), input_string.end(), [] (const char &x, const char &y) {
        return x == y and x == ' ';
    });

    input_string.erase(new_end, input_string.end());

    while (input_string[input_string.length() - 1] == ' ') {
        input_string.pop_back();
    }

    vector<string> splits;
    char delimiter = ' ';

    size_t i = 0;
    size_t pos = input_string.find(delimiter);

    while (pos != string::npos) {
        splits.push_back(input_string.substr(i, pos - i));

        i = pos + 1;
        pos = input_string.find(delimiter, i);
    }

    splits.push_back(input_string.substr(i, min(pos, input_string.length()) - i + 1));

    return splits;
}